decorator_allowed_groups/models/decorator.py

```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2021-Today: GRAP (<http://www.grap.coop/>)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api
+from odoo.exceptions import AccessError
+
+
+def allowed_groups(*group_xml_ids):
+    """ Return a decorator that specifies group(s)
+        to which the user must belong in order to perform
+        this function.
+        - if the user does not belong to any group,
+        an AccessError will be raised if the function is called.
+        - Also in the generated views, the according button
+        will be hidden if the user doesn't belong to the group(s).
+
+            @api.allowed_groups(
+                "purchase.group_purchase_manager",
+                "sale.group_sale_manager"
+            )
+            def my_secure_action(self):
+                pass
+    """
+
+    def decorator(method):
+        def secure_method(*args, **kwargs):
+            _self = args[0]
+            # Check if current user is member of correct groups
+            if any([
+                _self.env.user.has_group(group_xml_id)
+                for group_xml_id in group_xml_ids
+            ]):
```

Comment:
```suggestion
            if any(
                _self.env.user.has_group(group_xml_id)
                for group_xml_id in group_xml_ids
            ):
```

Otherwise the full list is built before any starts looking into it.

Reply:
You're right ! In fact, I replace by build in function that does exactly what I want. ``user_has_groups``.
(https://github.com/odoo/odoo/blob/12.0/odoo/models.py#L1204)